### PR TITLE
Write garmin specific ilinks as gpxx:rpt elements

### PR DESCRIPTION
**Pull request by lintondf**
The garmin GDB module already extracts and stores autoroute interlink (intermediate lat/lon points) data.  This pull request outputs this data when the "garminextension" GPX option is specified as gpxx:RoutePointExtension elements.

Test data and test driver added.

### CODE REVIEW

File: gpx.cc [hunk 1]
```diff
@@ -1289,7 +1289,10 @@ void
 GpxFormat::gpx_write_common_extensions(const Waypoint* waypointp, const gpx_point_type point_type) const
 {
   // gpx version we are writing is >= 1.1.
+  garmin_fs_t* gmsd = garmin_fs_t::find(waypointp);
+
   if ((opt_humminbirdext && (WAYPT_HAS(waypointp, depth) || WAYPT_HAS(waypointp, temperature))) ||
+      (opt_garminext && gmsd != NULL && gmsd->ilinks != NULL)  ||
```
Comment (tsteven4) [inline gpx.cc]: please check point_type.  One way to get ilinks with a different point type is using the transform filter, e.g. "-x transform,wpt=rte,del".  The check will prevent empty extensions elements from being written.

opt_garminext && gpxpt_route==point_type && ...

File: gpx.cc [hunk 2]
```diff
@@ -1325,7 +1328,22 @@ GpxFormat::gpx_write_common_extensions(const Waypoint* waypointp, const gpx_poin
         }
         break;
       case gpxpt_route:
-        /* we don't have any appropriate data for the children of gpxx:RoutePointExtension */
+        if (gmsd != NULL && gmsd->ilinks != NULL) {
+          writer->writeStartElement(QStringLiteral("gpxx:RoutePointExtension"));
+          garmin_ilink_t* link = gmsd->ilinks;
+          garmin_ilink_t* prior = NULL;
```
Comment (tsteven4) [inline gpx.cc]: unfortunate that the gdb seems to include duplicate links.

File: gpx.cc [hunk 3]
```diff
@@ -1289,7 +1289,10 @@ void
 GpxFormat::gpx_write_common_extensions(const Waypoint* waypointp, const gpx_point_type point_type) const
 {
   // gpx version we are writing is >= 1.1.
+  garmin_fs_t* gmsd = garmin_fs_t::find(waypointp);
```
Comment (tsteven4) [inline gpx.cc]: I am wondering about making this search conditional for efficiency in the normal case, i.e. when not writing garmin extensions, e.g.
  garmin_fs_t* gmsd = opt_garminext? garmin_fs_t::find(waypointp) : nullptr;

**Review by tsteven4 — CHANGES_REQUESTED** [review]
I am still chewing on this, but here is what I have so far.

please prefer c++ nullptr instead of c NULL macro. see https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-nullptr.html.  clang-tidy can do this for you.  If you don't have clang-tidy you can do it by hand, or if you allowed maintainers to edit (which is the default) I could push the clang-tidy changes to lintondf:gpxoutputilinks.
```
 rm gpx.o
 bear make gpx.o
 make clang-tidy RUN_CLANG_TIDY_FLAGS="-header-filter=.*\\\.h -checks=-*,modernize-use-nullptr -fix"
```

File: gpx.cc (hunk 3 above)
Comment (tsteven4) [reply gpx.cc]: Here is a way to only search for gmsd when using garmin extensions (as is done in waypt.cc).  Also checks point_type.

```
diff --git a/gpx.cc b/gpx.cc
index 915063c3..90d75613 100644
--- a/gpx.cc
+++ b/gpx.cc
@@ -1289,10 +1289,10 @@ void
 GpxFormat::gpx_write_common_extensions(const Waypoint* waypointp, const gpx_point_type point_type) const
 {
   // gpx version we are writing is >= 1.1.
-  garmin_fs_t* gmsd = garmin_fs_t::find(waypointp);
+  garmin_fs_t* gmsd = nullptr;
 
   if ((opt_humminbirdext && (WAYPT_HAS(waypointp, depth) || WAYPT_HAS(waypointp, temperature))) ||
-      (opt_garminext && gmsd != NULL && gmsd->ilinks != NULL)  ||
+      (opt_garminext && gpxpt_route==point_type && (gmsd = garmin_fs_t::find(waypointp)) && (gmsd->ilinks != nullptr))  ||
       (opt_garminext && gpxpt_waypoint==point_type && (WAYPT_HAS(waypointp, proximity) || WAYPT_HAS(waypointp, temperature) || WAYPT_HAS(waypointp, depth))) ||
       (opt_garminext && gpxpt_track==point_type && (WAYPT_HAS(waypointp, temperature) || WAYPT_HAS(waypointp, depth) || waypointp->heartrate != 0 || waypointp->cadence != 0))) {
     writer->writeStartElement(QStringLiteral("extensions"));
```

**Comment by lintondf** [comment]
I believe I responded to your requested changes

File: gpx.cc [hunk 4]
```diff
@@ -1325,7 +1328,22 @@ GpxFormat::gpx_write_common_extensions(const Waypoint* waypointp, const gpx_poin
...
+        if (gmsd != nullptr && gpxpt_route==point_type && gmsd->ilinks != nullptr) {
+          writer->writeStartElement(QStringLiteral("gpxx:RoutePointExtension"));
+          garmin_ilink_t* link = gmsd->ilinks;
+          garmin_ilink_t* prior = nullptr;  // GDB files sometime contain repeated point; omit them
+          while (link != nullptr) {
+            if (prior == nullptr || prior->lat != link->lat || prior->lon != link->lon) {
+              writer->writeStartElement(QStringLiteral("gpxx:rpt"));
+              writer->writeAttribute(QStringLiteral("lat"), toString(link->lat));
+              writer->writeAttribute(QStringLiteral("lon"), toString(link->lon));
+              writer->writeEndElement(); // "gpxx:RoutePointExtension"
```
Comment (tsteven4) [inline gpx.cc]: This comment has been erroneously copied, please delete it

**Comment by tsteven4** [comment]
Has the addition of the gpxx:rpts been of any use to you?  Is there some program that can interpret these?

**Comment by lintondf** [comment]
Yes, they provide detailed road shape point between the auto route waypoints.  The fields are present when you export GPX from BaseCamp.   Adding them to gpsbabel let’s me get everything I need from the GDB file.

**Comment by tsteven4** [comment]
> Yes, they provide detailed road shape point between the auto route waypoints. The fields are present when you export GPX from BaseCamp. Adding them to gpsbabel let’s me get everything I need from the GDB file.

When I read the gpsbabel generated gpx file gdb-sample-v3-ilinks.gpx into BC it doesn't seem to use the gpxx:rpt elements.  But perhaps you aren't reading the gpx into BC but using it in some other fashion.

**Comment by tsteven4** [comment]
@lintondf, are you good with the current behavior of reading the gpsbabel generated gpx file into basecamp?

> When I read the gpsbabel generated gpx file gdb-sample-v3-ilinks.gpx into BC it doesn't seem to use the gpxx:rpt elements. But perhaps you aren't reading the gpx into BC but using it in some other fashion.

**Comment by GPSBabelDeveloper** [comment]
I find this weird, but hand my approval to Steven.

**Comment by lintondf** [comment]
With the outstanding pull request code included it meets all my needs.

**Comment by tsteven4** [comment]
thanks

**Pull request merged**